src/pudl_archiver/depositors/depositor.py

```diff
@@ -0,0 +1,336 @@
+"""Implements generic interface for depositors."""
+
+import io
+import logging
+import re
+from abc import ABC, abstractmethod
+from contextlib import asynccontextmanager
+from dataclasses import dataclass
+from enum import Enum, auto
+from pathlib import Path
+from typing import BinaryIO
+
+import aiohttp
+from pydantic import BaseModel, ConfigDict, PrivateAttr
+
+from pudl_archiver.archivers.validate import RunSummary
+from pudl_archiver.frictionless import DataPackage, ResourceInfo
+from pudl_archiver.utils import RunSettings, Url
+
+logger = logging.getLogger(f"catalystcoop.{__name__}")
+
+
+class _UploadSpec(BaseModel):
+    """Defines an upload that will be done by ZenodoDepositionInterface."""
+
+    source: io.IOBase | Path
+    dest: str
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+
+class FileWrapper(io.BytesIO):
+    """Minimal wrapper around BytesIO to override close method to work around aiohttp."""
+
+    def __init__(self, content: bytes):
+        """Call base class __init__."""
+        super().__init__(content)
+
+    def close(self):
+        """Don't close file, so aiohttp can't unexpectedly close files."""
+        pass
+
+    def actually_close(self):
+        """Actually close the file for internal use."""
+        super().close()
+
+
+class DepositionAction(Enum):
+    """Enumerate types of changes which can be applied to deposition files."""
+
+    CREATE = (auto(),)
+    UPDATE = (auto(),)
+    DELETE = (auto(),)
```

Comment:
If we allow the action to be `NO_OP` also, we can stop using `| None` in the type signatures 🙂

Reply:
Good call, changed